### PR TITLE
[Operational Tool] Add --disable-address-validation flag to avoid spurious address errors

### PR DESCRIPTION
**Pull request by JoshLind**
## Motivation

This PR adds a flag `--disable-address-validation` to the commands in the op tool that update the validator config. It also disables address validation for any key rotation operations (consensus key, validator network key, and full node network key), as these operations are orthogonal to setting the network address. Without this PR, we currently get spurious address validation errors when trying to rotate keys (e.g., when my machine can't resolve an address locally, but that address is indeed valid in the deployment environment).

To achieve this, the PR offers several commits:
1. Add the --disable-address-validation flag to the tool, and disable address validation for all key rotation operations.
2. Add a missing smoke test for the full node network key rotation command (orthogonal, but needed).
3. Add a smoke test for the new flag, as well as update the test helper
4. Update auto-validate to use a checked add

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

All smoke tests pass locally, including the newly added ones. Moreover, running the tool locally now allows me to rotate the consensus key, whereas it previously fails (because my machine can't connect to the AWS address that's only available within the k8s deployment...)

Before this PR:
```
% ./libra-operational-tool rotate-consensus-key --config k8_config_file                             
{
  "Error": "Invalid arguments: validator address: Failed to resolve address '/dns4/val0-libra-validator-validator-lb/tcp/6180': failed to lookup address information: nodename nor servname provided, or not known"
}
```

After this PR:
```
% ./libra-operational-tool rotate-consensus-key --config k8_config_file
{
  "Result": "Executed"
}

```

## Related PRs

This PR is built on the auto validate PR: https://github.com/libra/libra/pull/6560

### CODE REVIEW

File: config/management/operational/src/account_resource.rs
Comment (gregnazario) [inline config/management/operational/src/account_resource.rs:134]: Btw, the build on top of the previous commits thing is kinda weird, but I do like you separated it out by which commits

File: config/management/operational/src/test_helper.rs
Comment (gregnazario) [inline config/management/operational/src/test_helper.rs:308]: I think we might want to either not validate the addresses, or allow it to be configurable.  Makes sense to me for this to be `true` because the operator key doesn't affect the network addresses (therefore if the addresses were wrong already, we shouldn't stop operator key rotation)

File: config/management/operational/src/test_helper.rs
Comment (gregnazario) [inline config/management/operational/src/test_helper.rs:294]: I think this is another instance where, basically we should only validate the addresses when someone specifically changes them?

File: config/management/src/validator_config.rs
Comment (gregnazario) [inline config/management/src/validator_config.rs:50]: Okay, so now that I've looked at this and I'm thinking about it.  We should probably not validate addresses when we rotate the keys.  The addresses were already put in previously, and rotating a key is not at the fault of the current operator.  Only when an operator puts in the non-key information, should it validate that it's correct.  Thoughts?

**Comment by mimoo** [comment]
for the #integer-overflow finding, I just suggest that you mark the whole crate as less risky via `#![allow(clippy::integer_arithmetic)]`

(or you can also do the right thing and use a checked operation, but I presume we don't care that much for operational toolings)

File: config/management/operational/src/account_resource.rs
Comment (JoshLind) [reply config/management/operational/src/account_resource.rs:134]: Thanks @gregnazario! Yeah, I was kind of hoping that would be reviewed first.. so we'd ultimately not need to worry about those commits :P

File: config/management/operational/src/test_helper.rs
Comment (JoshLind) [reply config/management/operational/src/test_helper.rs:308]: See my comment below.

File: config/management/operational/src/test_helper.rs
Comment (JoshLind) [reply config/management/operational/src/test_helper.rs:294]: See my comment below.

File: config/management/src/validator_config.rs
Comment (JoshLind) [reply config/management/src/validator_config.rs:50]: Yeah, I basically did this change with the mindset of making as few code changes as possible. Given that they all eventually call into the set-validator-config command, I figured we could just propagate the flag.  I was trying to avoid refactoring everything. However, it might make sense to consider this now. Let's hold off until the original has landed.

**Comment by JoshLind** [comment]
Thanks all, let's hold off until the base PR has landed first. Will return to this aftwards.

File: config/management/src/validator_config.rs
Comment (davidiw) [reply config/management/src/validator_config.rs:50]: actually, I would argue it is nice to have this validation... maybe it is over kill, I dunno, but if we have a flag to disable it... I think it is good.

File: config/management/operational/src/command.rs
Comment (davidiw) [inline config/management/operational/src/command.rs:121]: if you do this, you're going to have to update the scripts and the runbooks.....

Comment (davidiw) [reply config/management/operational/src/command.rs:121]: I think

**Comment by JoshLind** [comment]
Now that the auto validate PR has landed, I've updated this PR:
- I've disabled all network address validations for key rotations. As @gregnazario pointed out, they aren't that helpful. Moreover, failing on an invalid address when performing a key rotation is really confusing from a user perspective. Instead it's just the set-validator-config command in the op tool, and the validator-config command in the genesis tool that should do validation.
- I've updated the previously flagged potential integer overflow/underflow to use a checked add. If this happens, the tool should just panic. I don't want to ignore it for the whole crate because there might be other interesting cases ('cc @mimoo).

File: config/management/operational/src/command.rs
Comment (JoshLind) [reply config/management/operational/src/command.rs:121]: So there's two points to make here:
- The existing runbooks don't actually talk about this command at all. So we need to improve our runbooks.
- The command is actually "rotate-full-node-network-key" (note: RotateFullNodeNetworkKey -- "Node" has a capital, so this is currently a bug which wasn't detected because there was no smoke test for it). I don't mind which we use, so I chose the variant with the smallest number of code changes. Overall, it's probably worth pointing out that there's a lot of places in our codebase where we use "full node" and "fullnode". It's not great and it's kinda ugly. Perhaps we should discuss this with the networking folks and settle on something so we can clean up the codebase?

File: testsuite/smoke-test/src/operational_tooling.rs
Comment (davidiw) [inline testsuite/smoke-test/src/operational_tooling.rs:196]: I don't really think we need a smoke test for this, when we can get everything we need from an integration test.

**Review by davidiw — APPROVED** [review]
consider deleting the smoke test.

File: testsuite/smoke-test/src/operational_tooling.rs
Comment (JoshLind) [reply testsuite/smoke-test/src/operational_tooling.rs:196]: Yeah, I think that's fair, although we do check that transactions are being executed by the blockchain allowing us to set bad networking addresses. But for the other operations, we could check the tool works using integration/unit style testing. I'll leave this as-is for now though as it's simpler to check all this functionality in once place. Once we start to think about our testing principles we can take a look at all the tests to figure out where things should live.

**Comment by JoshLind** [comment]
Thanks, David! 😄 

/land

**Pull request closed without merging**
